File: examples/delivery/stream/agent.py
"""Compliance Checker Agent — stream delivery binding.

Delivery: stream  (GET /v1/agents/{address}/intents/stream)
The agent holds a persistent SSE connection to AXME.
When an intent arrives AXME pushes it over that connection immediately.

Run:
    export AXME_API_KEY=<your-api-key>
    export AXME_AGENT_ADDRESS=agent://<org>/<workspace>/compliance-checker-agent
    export AXME_BASE_URL=https://api.cloud.axme.ai   # or http://localhost:8000 for local dev

    python agent.py

Then in another terminal:
    axme scenarios apply examples/delivery/stream/scenario.json --watch
"""
from __future__ import annotations

import logging
import os
import sys
from typing import Any

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  [%(levelname)s]  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("compliance-checker")

# ---------------------------------------------------------------------------
# Configuration — all from environment, no hard-coded values
# ---------------------------------------------------------------------------

AXME_BASE_URL     = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY      = os.environ.get("AXME_API_KEY", "")
AXME_AGENT_ADDRESS = os.environ.get(
    "AXME_AGENT_ADDRESS",
    "compliance-checker-agent",  # bare name — resolved to full address by SDK
)

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")

# ---------------------------------------------------------------------------
# Business logic — real compliance rules, no stubs
# ---------------------------------------------------------------------------

_ALLOWED_CHANGE_WINDOW = {"Mon", "Tue", "Wed", "Thu"}

_RISK_LEVEL_REQUIRES_BACKUP = {"high", "critical"}

_BLOCKED_ENVIRONMENTS = {"production"}


def _check_compliance(payload: dict[str, Any]) -> tuple[bool, str]:
    """Return (passed, reason).  Validates a change request against compliance rules."""
    change_id   = payload.get("change_id", "?")
    environment = payload.get("environment", "")
    risk_level  = payload.get("risk_level", "low")
    backup_ok   = bool(payload.get("backup_confirmed", False))

    if environment in _BLOCKED_ENVIRONMENTS:
        return False, f"{change_id}: production changes require manual approval, not automated"

    if risk_level in _RISK_LEVEL_REQUIRES_BACKUP and not backup_ok:
        return False, f"{change_id}: risk_level={risk_level} requires backup_confirmed=true"

    return True, f"{change_id}: compliance checks passed (environment={environment}, risk={risk_level})"


# ---------------------------------------------------------------------------
# Agent loop
# ---------------------------------------------------------------------------

def handle_intent(client: AxmeClient, delivery: dict[str, Any]) -> None:
    intent_id = delivery.get("intent_id", "")
    if not intent_id:
        log.warning("received delivery without intent_id, skipping")
        return

    log.info("received intent %s", intent_id)

    try:
        intent = client.get_intent(intent_id)
    except Exception as exc:
        log.error("get_intent(%s) failed: %s", intent_id, exc)
        return

    # Skip intents not in actionable state — avoids reprocessing old/completed intents.
    # CREATED is included for step-intents dispatched by agent-core via workflow-commands,
    # which are immediately visible in the SSE stream before any DELIVERED transition.
    status = (intent.get("lifecycle_status") or intent.get("status") or "").upper()
    if status not in {"CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"}:
        log.debug("skipping intent %s with status %s", intent_id, status)
        return

    payload = intent.get("payload") or {}
    passed, reason = _check_compliance(payload)

    log.info("compliance result for %s: passed=%s  reason=%s", intent_id, passed, reason)

    try:
        if passed:
            client.resume_intent(
                intent_id,
                {
                    "action":  "complete",
                    "passed":  True,
                    "reason":  reason,
                    "checked_fields": ["environment", "risk_level", "backup_confirmed"],
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        else:
            client.resume_intent(
                intent_id,
                {
                    "action": "fail",
                    "passed": False,
                    "reason": reason,
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        log.info("resumed intent %s  (passed=%s)", intent_id, passed)
    except Exception as exc:
        log.error("resume_intent(%s) failed: %s", intent_id, exc)


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))

    log.info("compliance-checker-agent starting  address=%s  binding=stream", AXME_AGENT_ADDRESS)
    log.info("listening on %s ...", AXME_BASE_URL)

    try:
        for delivery in client.listen(AXME_AGENT_ADDRESS, timeout_seconds=None):
            handle_intent(client, delivery)
    except KeyboardInterrupt:
        log.info("shutting down")


if __name__ == "__main__":
    main()

File: examples/delivery/stream/go/agent.go
// Compliance Checker Agent — stream delivery binding (Go).
//
// Run:
//
//	export AXME_API_KEY=<your-api-key>
//	export AXME_AGENT_ADDRESS=compliance-checker-agent
//	go run examples/delivery/stream/agent.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strings"

	"github.com/AxmeAI/axme-sdk-go/axme"
)

var (
	baseURL      = envOr("AXME_BASE_URL", "https://api.cloud.axme.ai")
	apiKey       = os.Getenv("AXME_API_KEY")
	agentAddress = envOr("AXME_AGENT_ADDRESS", "compliance-checker-agent")
)

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// --- Business logic ---

var (
	blockedEnvironments  = map[string]bool{"production": true}
	riskRequiresBackup   = map[string]bool{"high": true, "critical": true}
)

func checkCompliance(payload map[string]any) (bool, string) {
	changeID := str(payload, "change_id", "?")
	environment := str(payload, "environment", "")
	riskLevel := str(payload, "risk_level", "low")
	backupOk, _ := payload["backup_confirmed"].(bool)

	if blockedEnvironments[environment] {
		return false, fmt.Sprintf("%s: production changes require manual approval, not automated", changeID)
	}
	if riskRequiresBackup[riskLevel] && !backupOk {
		return false, fmt.Sprintf("%s: risk_level=%s requires backup_confirmed=true", changeID, riskLevel)
	}
	return true, fmt.Sprintf("%s: compliance checks passed (environment=%s, risk=%s)", changeID, environment, riskLevel)
}

func str(m map[string]any, key, def string) string {
	if v, ok := m[key].(string); ok && v != "" {
		return v
	}
	return def
}

func main() {
	if apiKey == "" {
		log.Fatal("AXME_API_KEY is required")
	}

	client, err := axme.NewClient(axme.ClientConfig{APIKey: apiKey, BaseURL: baseURL})
	if err != nil {
		log.Fatalf("client init failed: %v", err)
	}

	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
	defer cancel()

	log.Printf("compliance-checker-agent starting  address=%s  binding=stream", agentAddress)

	intents, errs := client.Listen(ctx, agentAddress, axme.ListenOptions{})
	for {
		select {
		case delivery, ok := <-intents:
			if !ok {
				return
			}
			intentID := str(delivery, "intent_id", "")
			if intentID == "" {
				continue
			}
			log.Printf("received intent %s", intentID)
			handleIntent(ctx, client, intentID)
		case err, ok := <-errs:
			if !ok {
				return
			}
			log.Printf("stream error: %v", err)
		case <-ctx.Done():
			log.Println("shutting down")
			return
		}
	}
}

func handleIntent(ctx context.Context, client *axme.Client, intentID string) {
	resp, err := client.GetIntent(ctx, intentID, axme.RequestOptions{})
	if err != nil {
		log.Printf("get_intent(%s) failed: %v", intentID, err)
		return
	}
	intent, _ := resp["intent"].(map[string]any)
	if intent == nil {
		intent = resp
	}

	status := strings.ToUpper(str(intent, "lifecycle_status", str(intent, "status", "")))
	actionable := map[string]bool{"CREATED": true, "DELIVERED": true, "ACKNOWLEDGED": true, "IN_PROGRESS": true, "WAITING": true}
	if !actionable[status] {
		return
	}

	rawPayload, _ := intent["payload"].(map[string]any)
	if rawPayload == nil {
		rawPayload = map[string]any{}
	}
	effectivePayload, _ := rawPayload["parent_payload"].(map[string]any)
	if effectivePayload == nil {
		effectivePayload = rawPayload
	}

	passed, reason := checkCompliance(effectivePayload)
	log.Printf("compliance result for %s: passed=%v  reason=%s", intentID, passed, reason)

	var resumePayload map[string]any
	if passed {
		resumePayload = map[string]any{
			"action": "complete", "passed": true, "reason": reason,
			"checked_fields": []string{"environment", "risk_level", "backup_confirmed"},
		}
	} else {
		resumePayload = map[string]any{"action": "fail", "passed": false, "reason": reason}
	}

	_, err = client.ResumeIntent(ctx, intentID, resumePayload, axme.RequestOptions{OwnerAgent: agentAddress})
	if err != nil {
		log.Printf("resume_intent(%s) failed: %v", intentID, err)
		return
	}
	log.Printf("resumed intent %s  (passed=%v)", intentID, passed)
}
